src/python/gem5/utils/multisim/multisim.py

```diff
@@ -0,0 +1,245 @@
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""This module contains the gem5 MultiSim framework. The gem5 MultiSim
+work functions by allowing a user to specify multiple simulations to run from
+a single gem5 config script. The MuliSim framework will then run these
+simulations in parallel using the Python multiprocessing library.
+
+The framework works by having the user add the simulators in a configuration
+via the `add_simulator` function, each with a unique, user specified, id. This
+adds the different simulations to run in parallel to a global list. The
+MultiSim framework then uses the Python multiprocessing library to run the
+simulations in parallel by loading the config script as a module in each child
+process and then selecting the simulation to run via the iid in the global set
+of scheduled simulators jobs to run.
+The only difference between the child processes is the id of the simulator.
+
+Important notes
+---------------
+
+1. You cannot load/instantiate the simulators in the main process. You cannot
+even load the config script (i.e., `import config_script`). This means the
+config script is passed as a string referencing the config script as a module.
+This script is then passed to the child processes to load.
+
+2. The config script cannot accept parameters. It must be parameterless.
+"""
+
+import importlib
+import multiprocessing
+from pathlib import Path
+from typing import (
+    Optional,
+    Set,
+)
+
+# A global variable which __main__.py flips to `True` when multisim is run as
+# an executable module.
+module_run = False
+
+_multi_sim: Set["Simulator"] = set()
```

Comment:
Just curious (and related to the previous version with the `hash` implementation)... is it possible to have two simulators that are *exactly* the same and run both simulations? I actually want this feature, because I may have some pseudo random number which makes the simulations behave differently.

Reply:
I've changed this back to List just because it makes more sense for how I have things setup now, however, if it were a set, it would still be possible. Simulator objects can be exactly the same. The Python hash and equals operator defaults to the object pointer, not anything to do with the contents.

Reply:
Thanks!